sklearn/decomposition/nmf.py

```diff
@@ -78,19 +81,156 @@ def _safe_compute_error(X, W, H):
     return error
 
 
-def _check_string_param(sparseness, solver):
+def beta_divergence(X, W, H, beta):
+    """Compute the beta-divergence of X and dot(W, H).
+
+    If beta == 2, this is the Frobenius squared norm
+    If beta == 1, this is the generalized Kullback-Leibler divergence
+    If beta == 0, this is the Itakura-Saito divergence
+    Else, this is the general beta-divergence.
+    """
+    beta = _beta_loss_to_float(beta)
+
+    # The method can be called with scalars
+    if isinstance(X, numbers.Number):
+        W = np.array([[W]])
+        H = np.array([[H]])
+        X = np.array([[X]])
+
+    # Frobenius norm
+    if beta == 2:
+        # Avoid the creation of the dense np.dot(W, H) if X is sparse.
+        if sp.issparse(X):
+            norm_X = np.dot(X.data, X.data)
+            norm_WH = trace_dot(np.dot(np.dot(W.T, W), H), H)
+            cross_prod = trace_dot((X * H.T), W)
+            res = (norm_X + norm_WH - 2. * cross_prod) / 2.
+        else:
+            res = squared_norm(X - np.dot(W, H)) / 2.
+        return res
+
+    if sp.issparse(X):
+        # compute np.dot(W, H) only where X is nonzero
+        WH_data = _special_sparse_dot(W, H, X).data
+        X_data = X.data
+    else:
+        WH = fast_dot(W, H)
+        WH_data = WH.ravel()
+        X_data = X.ravel()
+
+    # do not affect the zeros: here 0 ** (-1) = 0 and not infinity
+    WH_data = WH_data[X_data != 0]
+    X_data = X_data[X_data != 0]
+
+    # used to avoid division by zero
+    WH_data[WH_data == 0] = EPSILSON
+
+    # generalized Kullback-Leibler divergence
+    if beta == 1:
+        # fast and memory efficient computation of np.sum(np.dot(W, H))
+        sum_WH = np.dot(np.sum(W, axis=0), np.sum(H, axis=1))
+        # computes np.sum(X * log(X / WH)) only where X is nonzero
+        div = X_data / WH_data
+        res = np.dot(X_data, np.log(div))
+        # add full np.sum(np.dot(W, H)) - np.sum(X)
+        res += sum_WH - X_data.sum()
+
+    # Itakura-Saito divergence
+    elif beta == 0:
+        div = X_data / WH_data
+        res = np.sum(div) - np.product(X.shape) - np.sum(np.log(div))
+
+    # beta-divergence, beta not in (0, 1, 2)
+    else:
+        if sp.issparse(X):
+            # slow loop, but memory efficient computation of :
+            # np.sum(np.dot(W, H) ** beta)
+            sum_WH_beta = 0
+            for i in range(X.shape[1]):
+                sum_WH_beta += np.sum(fast_dot(W, H[:, i]) ** beta)
```

Comment:
This is a temporary fix so as to be memory efficient in the sparse case.
(We cannot compute directly `np.dot(W, H)`).

We could have a cython code for it, yet the use case is rather limited (sparse X, and beta not in [0, 1, 2] which are the most useful cases), so I propose to merge it as it is, and improve it in a separate PR.